### PR TITLE
Nodes Recovery::recover Azure Infrastructures

### PR DESCRIPTION
**Pull request by jrochas**

### CODE REVIEW

File: infrastructures/infrastructure-aws-ec2/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AWSEC2Infrastructure.java [hunk 1]
```diff
@@ -228,7 +214,11 @@ public void acquireNode() {
             List<String> scripts = Lists.newArrayList(this.downloadCommand,
                                                       "nohup " + generateDefaultStartNodeCommand(instanceId) + "  &");
 
-            connectorIaasController.executeScript(getInfrastructureId(), instanceId, scripts);
+            try {
+                connectorIaasController.executeScript(getInfrastructureId(), instanceId, scripts);
+            } catch (ScriptNotExecutedException e) {
+                logger.info("Script not executed for instance " + instanceId);
```
Comment (marcocast) [inline infrastructures/infrastructure-aws-ec2/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AWSEC2Infrastructure.java]: Why info level? also we can print the Exception in the log.

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java [hunk 2]
```diff
@@ -301,29 +295,66 @@ public void acquireNode() {
...
+                                                                        vmPassword,
+                                                                        vmPublicKey,
+                                                                        vmSizeType,
+                                                                        resourceGroup,
+                                                                        region,
+                                                                        privateNetworkCIDR,
+                                                                        staticPublicIP);
+            LOGGER.info("Instances ids created or retrieved : " + instancesIds);
+        } else {
+            // if the infrastructure was already created, then wee need to
```
Comment (marcocast) [inline infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: we need to...

File: infrastructures/infrastructure-common/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/ScriptNotExecutedException.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ *
+ * A checked exception to signal that the script was not executed when the
+ * provider instance was acquired. Wrap the underlying exception.
+ */
+public class ScriptNotExecutedException extends Exception {
+
+    private Exception e;
+
+    public ScriptNotExecutedException(Exception e) {
+        this.e = e;
```
Comment (marcocast) [inline infrastructures/infrastructure-common/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/ScriptNotExecutedException.java]: just do 
super (e)
and u can remove all the rest including getUnderlyingException

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java [hunk 4]
```diff
@@ -301,29 +295,66 @@ public void acquireNode() {
...
+        // execute script on instances to deploy or redeploy nodes on them
         for (String instanceId : instancesIds) {
-
             String fullScript = generateScriptFromInstanceId(instanceId);
-            connectorIaasController.executeScript(getInfrastructureId(), instanceId, Lists.newArrayList(fullScript));
+            try {
+                connectorIaasController.executeScript(getInfrastructureId(),
+                                                      instanceId,
+                                                      Lists.newArrayList(fullScript));
+            } catch (Exception e) {
```
Comment (vinseon) [inline infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: Why not catching the new `ScriptNotExecutedException`?

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java [hunk 5]
```diff
@@ -301,29 +295,66 @@ public void acquireNode() {
...
+                    acquireNode();
+                    break;
+                } else {
+                    LOGGER.info("Script execution failed and cannot be handled, abandoning instance " + instanceId);
+                }
+            } finally {
+                // in all cases, we must remove the instance from the free
+                // instance map as we tried everything to deploy nodes on it
+                removeFreeInstanceFromMap(instanceId);
+            }
```
Comment (vinseon) [inline infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: All this catch/finally logic that redeploy the node couldn't be implemented the same way in OpenStack and VMWare?

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java [hunk 6]
```diff
@@ -301,29 +295,66 @@ public void acquireNode() {
...
+                                                      instanceId,
+                                                      Lists.newArrayList(fullScript));
+            } catch (Exception e) {
+                // if we cannot execute the script although the infrastructure
+                // was already deployed, then it means that the Azure
+                // instances are probably dead, so we will attempt a
+                // redeployment from scratch
+                if (instanceIdSaved) {
+                    LOGGER.info("Saved instance: " + instanceId + " does not exist anymore. Recreating all instances.");
+                    clearFreeInstancesMap();
```
Comment (vinseon) [inline infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: Here you clear all free instances, why not just removeFreeInstanceFromMap(instanceId)?

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java (hunk 2 above)
Comment (vinseon) [inline infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: probably weed in mind :-D

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java (hunk 4 above)
Comment (jrochas) [reply infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: yes

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java (hunk 6 above)
Comment (jrochas) [reply infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: I'll add a comment to explain the process

File: infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java (hunk 5 above)
Comment (jrochas) [reply infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java]: This is planed in longer term :) but first for EC2

**Pull request merged**